### ISSUE TITLE
Broken toplevel printing for types and modules (rec/nonrec not shown correctly)

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#7453]
**Original bug ID:** 7453
**Reporter:** @yallop
**Assigned to:** @diml
**Status:** assigned (set by @diml on 2017-01-13T17:39:51Z)
**Resolution:** open
**Priority:** normal
**Severity:** minor
**Category:** toplevel
**Monitored by:** @gasche @hcarty

## Bug description

Recursive type definitions are incorrectly displayed as explicitly non-recursive definitions:

    # #show list;;
    type nonrec 'a list = [] | (::) of 'a * 'a list
    # type t = T of t;;
    type t = T of t
    # #show t;;
    type nonrec t = T of t

And recursive module definitions are incorrectly displayed as implicitly non-recursive definitions:

    # module rec M : sig type t val x : M.t end = struct type t = int let x = 0 end;;
    module rec M : sig type t val x : M.t end
    # #show_module M;;
    module M : sig type t val x : M.t end

**Comment by vicuna** [comment]
**Comment author:** @diml

I had a look at this. I wrote two small functions to check if a type/module is recursive and select the rec_status field in the implementation of "show_XXX" according to this. The branch is here:

  https://github.com/diml/ocaml/tree/fix-toplevel-type-and-module-printing

It seems to work, there is just a bug when calling #show_module multiple times that I haven't fixed yet.

Another approach would be to record the recursive status inside the type/module declaration, so that the output of #show matches what the user wrote. Currently you get nonrec if the type is not in fact recursive:

```
# type t = int;;
type t = int
# #show t;;
type nonrec t = int
```

**Comment by vicuna** [comment]
**Comment author:** @yallop

Thanks for looking at this.

I don't think that preserving what the user wrote is especially worthwhile; the important thing is that what is printed is as close to correct as possible.  If it's not too difficult, I think the best scheme would be:

   * default to printing 'module' and switch to 'module rec' where the module type mentions the module name.  I think this is what you've implemented already.

   * default to printing 'type' and switch to 'type nonrec' where necessary (i.e. where the type definition would mean something different if you left out the 'nonrec').  This is a bit different to your current scheme.

For actually-recursive types there's no difference between your current scheme and the proposal above.  And there's no difference for types that actually need 'nonrec', either.  The only difference is for non-recursive types that can be written equivalently either with or without 'nonrec', such as 'type t = int', and it's best if those are printed as simply as possible.

**Comment by vicuna** [comment]
**Comment author:** voglerr

Is there something blocking this?
Had a bit of a problem explaining this to students...

> default to printing 'type' and switch to 'type nonrec' where necessary
> best if those are printed as simply as possible

Sounds perfect.

**Comment by yallop** [comment]
The code in the problem report now works in `trunk`:

```
# #show list;;
type 'a list = [] | (::) of 'a * 'a list
# type t = T of t;;
type t = T of t
# #show t;;
type t = T of t
```

But printing types defined using `nonrec` doesn't use `nonrec`, so there's still a bug:
```
# type nonrec list = T of int list;;
type nonrec list = T of int list
# #show list;;
type list = T of int list
```

**Comment by gasche** [comment]
I added a known-bug test for these faulty (non)use of `rec` by the #show command in https://github.com/ocaml/ocaml/commit/f7ec223df9c9c4871040251fb1f558e0876d379a.

One slightly strange behavior of the test, unrelated to the present discussion (it is not with `#show` but with standard toplevel printing) is the scoping-incorrect use of a `M/2` stamp in the second declaration below (cc @Octachron ; I believe this is a known and tolerated misbehavior but I couldn't find an issue to reference).

```ocaml
# module M : sig end = struct end;;
module M : sig  end
# module rec M : sig type t = Loop of M.t end = struct type t = Loop of M.t end;;
module rec M : sig type t = Loop of M/2.t end
```

**Comment by Octachron** [comment]
The printing issue is an instance of #8917. There is a fix for the bug, #8929, waiting for a review too.

**Comment by yallop** [comment]
This is not quite fixed yet: @gasche's [`broken_rec_in_show.ml`](https://github.com/ocaml/ocaml/blob/9ffd97f3eec0c45ead5e395a298b4a4a54fe26c1/testsuite/tests/tool-toplevel/known-bugs/broken_rec_in_show.ml) tests still fail in `trunk`.

**Comment by Octachron** [comment]
All examples are fixed by #10416 (and this time without fatal errors when `-short-paths` is enabled). There are still some case that are not perfectly printed, when the displayed fragment of a group of mutually recursive definitions is not self-recursive. But that only happens for unprintable type definitions (e.g.
```ocaml
type t;;
type u = [ `U of t]
type t = A of [ u | `V ] | B of s
and s = C of t;;
#show t;;
```
) and recursive modules.

**Comment by yallop** [comment]
Thank you, @Octachron!
